File: crypto/src/main/java/io/oversec/one/crypto/encoding/XCoderFactory.java
package io.oversec.one.crypto.encoding;

import android.content.Context;
import io.oversec.one.crypto.proto.Outer;

import java.util.ArrayList;
import java.util.List;


public class XCoderFactory {
    private static XCoderFactory INSTANCE;

    private List<IXCoder> ALL = new ArrayList<>();

    public final Base64XCoder _Base64XCoder;
    public final ZeroWidthXCoder _ZeroWidthXCoder;
    public final AsciiArmouredGpgXCoder _AsciiArmouredGpgXCoder;


    public static synchronized XCoderFactory getInstance(Context ctx) {
        if (INSTANCE == null) {
            INSTANCE = new XCoderFactory(ctx);
        }
        return INSTANCE;
    }

    private XCoderFactory(Context context) {
        _Base64XCoder = new Base64XCoder(context);
        _ZeroWidthXCoder = new ZeroWidthXCoder(context);
        _AsciiArmouredGpgXCoder = new AsciiArmouredGpgXCoder(context);
        add(_Base64XCoder);
        add(_ZeroWidthXCoder);
        add(_AsciiArmouredGpgXCoder);
    }

    private void add(IXCoder coder) {
        ALL.add(coder);
    }

    public synchronized Outer.Msg decode(String s) {
        for (IXCoder coder : ALL) {
            try {
                Outer.Msg m = coder.decode(s);
                if (m != null) {
                    return m;
                }
            } catch (Exception e) {
                //
            }
        }
        return null;
    }

    public synchronized String getEncodingInfo(String s) {
        for (IXCoder coder : ALL) {
            try {
                Outer.Msg msg = coder.decode(s);
                if (msg != null) {
                    return msg.getMsgDataCase().name() + " (" + coder.getLabel(null) + ")";
                }
            } catch (Exception e) {
                //
            }
        }
        return "N/A";

    }

    public boolean isEncodingCorrupt(String s) {
        for (IXCoder coder : ALL) {
            try {
                coder.decode(s);
            } catch (Exception e) {
                e.printStackTrace();
                return true;
            }
        }
        return false;
    }
}

File: crypto/src/main/java/io/oversec/one/crypto/encoding/XCoderFactory.kt
package io.oversec.one.crypto.encoding

import android.annotation.SuppressLint
import android.content.Context
import io.oversec.one.crypto.proto.Outer
import java.util.ArrayList

class XCoderFactory private constructor(context: Context) {

    private val ALL = ArrayList<IXCoder>()

    val base64XCoder = Base64XCoder(context).also { add(it) }
    val zeroWidthXCoder = ZeroWidthXCoder(context).also { add(it) }
    val asciiArmouredGpgXCoder = AsciiArmouredGpgXCoder(context).also { add(it) }


    private fun add(coder: IXCoder) {
        ALL.add(coder)
    }

    @Synchronized
    fun decode(s: String): Outer.Msg? {
        for (coder in ALL) {
            try {
                val m = coder.decode(s)
                if (m != null) {
                    return m
                }
            } catch (e: Exception) {
                //
            }

        }
        return null
    }

    @Synchronized
    fun getEncodingInfo(s: String): String {
        for (coder in ALL) {
            try {
                val msg = coder.decode(s)
                if (msg != null) {
                    return msg.msgDataCase.name + " (" + coder.getLabel(null) + ")"
                }
            } catch (e: Exception) {
                //
            }

        }
        return "N/A"

    }

    fun isEncodingCorrupt(s: String): Boolean {
        for (coder in ALL) {
            try {
                coder.decode(s)
            } catch (e: Exception) {
                e.printStackTrace()
                return true
            }
        }
        return false
    }

    companion object {
        @SuppressLint("StaticFieldLeak") // note that we're storing *Application*context
        @Volatile
        private var INSTANCE: XCoderFactory? = null

        fun getInstance(ctx: Context): XCoderFactory =
            INSTANCE ?: synchronized(this) {
                INSTANCE ?: XCoderFactory(ctx.applicationContext).also { INSTANCE = it }
            }
    }
}
